### PR TITLE
Set up reusable CI workflow and add building doctests

### PR DESCRIPTION
**Pull request by ian-h-chamberlain**
Part of #106 

This does a couple things:
* Makes sure we are actually building doctests now
* Switch to using https://github.com/actions-rust-lang/setup-rust-toolchain so we get nice inline warnings and caching
* Break out some common logic into a reusable composite action. I hope this can also be used for other crates going forward.

- [x] TODO update the git hash once we merge https://github.com/rust3ds/cargo-3ds/pull/32

### CODE REVIEW

**Review by Meziu — APPROVED** [review]
This is great! Thanks for setting up the CI this way. The more automated checks we have, the easier it is to maintain the repo 👍

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -68,5 +53,28 @@ jobs:
...
+    steps:
+      - name: Checkout branch
+        uses: actions/checkout@v2
+
+      - uses: ./.github/actions/setup
+        with:
+          toolchain: ${{ matrix.toolchain }}
+
+      - name: Hide duplicated warnings from lint job
+        run: echo "::remove-matcher owner=clippy::"
```
Comment (AzureMarker) [inline .github/workflows/ci.yml:74]: Is this needed for the doc test pipeline?

**Review by AzureMarker — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ian-h-chamberlain) [reply .github/workflows/ci.yml:74]: yeah, the matcher is added during setup by `actions-rust-lang/setup-rust-toolchain` and it's called `clippy` even though it also encompasses normal `rustc` diagnostics. In practice that means any compiler warnings would get duplicated here when the library is built by `cargo 3ds test --doc`, without this `::remove-matcher` in place.

File: .github/actions/setup/action.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      run: sudo apt-get update && sudo apt-get install -y build-essential
+
+    - name: Install cargo-3ds
+      uses: actions-rs/cargo@v1
+      with:
+        command: install
+        # TODO: this should probably just be a released version from crates.io
+        # once cargo-3ds gets published somewhere...
+        #
+        # also switch to master hash once feature/doctests gets merged
```
Comment (Meziu) [inline .github/actions/setup/action.yml]:
```suggestion
```

File: .github/actions/setup/action.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        args: >-
+          --git https://github.com/rust3ds/cargo-3ds
+          --rev 5a06e64307499ffe681453e7d88d7ee986054eb7
```
Comment (Meziu) [inline .github/actions/setup/action.yml]:
```suggestion
          --rev 78a652fdfb01e2614a792d1a56b10c980ee1dae9
```

**Review by Meziu — APPROVED** [review]
I’ve made the suggestions to update this PR and merge it.

**Pull request merged**